Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    for (auto key : iconsObject.keys()) {
+        // Filter the values to only include numeric strings representing sizes. This will exclude non-numeric string
+        // values such as "any", "color_schemes", and any other strings that cannot be converted to a positive integer.
+        auto integerValue = parseInteger<size_t>(key);
+        if (integerValue && integerValue > 0)
+            sizeValues.append(*integerValue);
+    }
+
+    if (!sizeValues.size())
+        return 0;
+
+    std::sort(sizeValues.begin(), sizeValues.end());
+
+    size_t bestSize = 0;
+    for (auto size : sizeValues) {
+        bestSize = size;
+        if (bestSize >= idealPixelSize)
+            break;
+    }
+
+    return bestSize;
+}
+
+String WebExtension::pathForBestImageInIconsDictionary(const JSON::Object& iconsObject, size_t idealPixelSize)
+{
+    size_t bestSize = bestSizeInIconsDictionary(iconsObject, idealPixelSize);
+    if (!bestSize)
+        return nullString();
+
+#if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
+    if (bestSize == std::numeric_limits<size_t>::max())
+        return iconsObject.getString(anyManifestKey);
+#endif
+
+    return iconsObject.getString(String::number(bestSize));
+}
+
+RefPtr<WebCore::Icon> WebExtension::bestImageForIconsDictionaryManifestKey(const JSON::Object& object, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error error, const String& customLocalizedDescription)
+{
+    auto currentScales = availableScreenScales();
+    if (auto cachedScales = cacheLocation.getOptional("scales"_s)) {
+        auto scales = std::get_if<Vector<double>>(&*cachedScales);
+        if (!scales || *scales != currentScales)
+            cacheLocation.set("scales"_s, IconVariant(currentScales));
+    } else
+        cacheLocation.set("scales"_s, IconVariant(currentScales));
+
+    auto cacheKey = idealSize.toJSONString();
+    if (auto cacheResult = cacheLocation.getOptional(cacheKey)) {
+        if (std::holds_alternative<RefPtr<WebCore::Icon>>(*cacheResult))
+            return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
+    }
+
+    RefPtr iconObject = object.getObject(manifestKey);
+    if (!iconObject)
+        return nullptr;
+
+    RefPtr result = bestImageInIconsDictionary(iconObject, idealSize, [&](Ref<API::Error> error) {
+        recordError(error);
+    });
+
+    if (!result) {
+        if (iconObject->size()) {
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
+            recordError(createError(error, customLocalizedDescription));
+        } else if (!iconObject->size() || object.getValue(manifestKey)) {
+            // Record an error if the key had dictionary that was empty, or the key had a value of the wrong type.
+            recordError(createError(error));
+        }
+
+        return nullptr;
+    }
+
+    cacheLocation.set(cacheKey, IconVariant(result));
```

Comment:
Is `IconVariant()` needed? I'm not sure if `std::variant` can be created with on of the alternatives. But this would be better if it works (applies to all other places that do this):
```suggestion
    cacheLocation.set(cacheKey, result);
```

Reply:
When I tried it without specifying IconVariant, it wouldn't allow me to create an std::variant from the type. So as far as I can tell, it is needed.